internal/manifest/manifest.go

```diff
@@ -0,0 +1,245 @@
+)
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	jsonwallDB, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+
+	manifest = &Manifest{db: jsonwallDB}
+	err = validate(manifest)
+	if err != nil {
+		return nil, err
+	}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePath(prefix string, f func(Path) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.IteratePrefix(map[string]string{"kind": "path", "path": prefix})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var path Path
+		err := iter.Get(&path)
+		if err != nil {
+			return err
+		}
+		err = f(path)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (manifest *Manifest) IteratePkgs(f func(Package) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.Iterate(map[string]string{"kind": "package"})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var pkg Package
+		err := iter.Get(&pkg)
+		if err != nil {
+			return err
+		}
+		err = f(pkg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (manifest *Manifest) IterateSlices(pkgName string, f func(Slice) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.IteratePrefix(map[string]string{"kind": "slice", "name": pkgName})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var slice Slice
+		err := iter.Get(&slice)
+		if err != nil {
+			return err
+		}
+		err = f(slice)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validate(manifest *Manifest) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf(`invalid manifest: %s`, err)
```

Comment:
There are backticks being used around the file in places that don't need them. Let's please stick to the double quotes unless it makes a difference.

Reply:
I tend to use backticks everywhere directly for consistency, but changed to double quotes where it makes sense.

Reply:
Let's please respect the convention of the code itself first.